File: src/games/pinclimb.c
#include <math.h>
#include <stdbool.h>
#include <stddef.h>

#include "cglp.h"

char *title = "PIN CLIMB";
char *description = "[Hold] Stretch";

#define CS char characters[][CHARACTER_HEIGHT][CHARACTER_WIDTH + 1]
CS = {};
int charactersCount = 0;

Options options = {
    .viewSizeX = 100, .viewSizeY = 100, .soundSeed = 1, .isDarkColor = false};

typedef struct {
  Vector pos;
  bool isAlive;
} Pin;

typedef struct {
  float angle;
  float length;
  Pin *pin;
} Cord;

Cord cord;
#define MAX_PIN_COUNT 32
Pin pins[MAX_PIN_COUNT];
int pinIndex;
float nextPinDist;
#define CORD_LENGTH 7

void addPin(float x, float y) {
  ASSIGN_ARRAY_ITEM(pins, pinIndex, Pin, p);
  p->pos.x = x;
  p->pos.y = y;
  p->isAlive = true;
  pinIndex = wrap(pinIndex + 1, 0, MAX_PIN_COUNT);
}

void update() {
  if (!ticks) {
    INIT_UNALIVED_ARRAY(pins);
    pinIndex = 0;
    addPin(50, 5);
    nextPinDist = 5;
    cord.angle = 0;
    cord.length = CORD_LENGTH;
    cord.pin = &pins[0];
    barCenterPosRatio = 0;
  }
  float scr = difficulty * 0.02;
  if (cord.pin->pos.y < 80) {
    scr += (80 - cord.pin->pos.y) * 0.1;
  }
  if (input.isJustPressed) {
    play(SELECT);
  }
  if (input.isPressed) {
    cord.length += difficulty;
  } else {
    cord.length += (CORD_LENGTH - cord.length) * 0.1;
  }
  cord.angle += difficulty * 0.05;
  bar(cord.pin->pos.x, cord.pin->pos.y, cord.length, cord.angle);
  if (cord.pin->pos.y > 98) {
    play(EXPLOSION);
    end();
  }
  Pin *nextPin = NULL;
  FOR_EACH(pins, i) {
    ASSIGN_ARRAY_ITEM(pins, i, Pin, p);
    SKIP_IS_NOT_ALIVE(p);
    p->pos.y += scr;
    if (box(p->pos.x, p->pos.y, 3, 3).isColliding.rect[BLACK] &&
        p != cord.pin) {
      nextPin = p;
    }
    p->isAlive = p->pos.y <= 102;
  }
  if (nextPin != NULL) {
    play(POWER_UP);
    addScore(ceil(distanceTo(&cord.pin->pos, nextPin->pos.x, nextPin->pos.y)),
             nextPin->pos.x, nextPin->pos.y);
    cord.pin = nextPin;
    cord.length = CORD_LENGTH;
  }
  nextPinDist -= scr;
  while (nextPinDist < 0) {
    addPin(rnd(10, 90), -2 - nextPinDist);
    nextPinDist += rnd(5, 15);
  }
}

File: convert_prompt/knowledge/pinclimb.js
// Write the game name to 'title'.
title = "PIN CLIMB";

// 'description' is displayed on the title screen.
description = `
[Hold] Stretch
`;

// User-defined characters can be written here.
characters = [];

// Configure game options.
options = {
  isPlayingBgm: true,
  isReplayEnabled: true,
  // If you want to play a different BGM or SE,
  // you can try changing the 'seed' value.
  seed: 400,
};

// (Optional) Defining the types of variables is useful for
// code completion and error detection.
/** @type {{angle: number, length: number, pin: Vector}} */
let bar;
/** @type {Vector[]} */
let pins;
let nextPinDist;
const barLength = 7;

// 'update()' is called every frame (60 times per second).
function update() {
  // 'ticks' counts the number of frames from the start of the game.
  if (!ticks) {
    // Initialize the game state here. (ticks === 0)
    pins = [vec(50, 0)]; // 'vec()' creates a 2d vector instance.
    nextPinDist = 10;
    bar = { angle: 0, length: barLength, pin: pins[0] };
  }
  // 'difficulty' represents the difficulty of the game.
  // The value of this variable is 1 at the beginning of the game and
  // increases by 1 every minute.
  let scr = difficulty * 0.02;
  if (bar.pin.y < 80) {
    scr += (80 - bar.pin.y) * 0.1;
  }
  // 'input.isJustPressed' is set to true the moment the button is pressed.
  if (input.isJustPressed) {
    // 'play()' plays the SE.
    play("select");
  }
  // 'input.isPressed' is set to true while the button is pressed.
  if (input.isPressed) {
    bar.length += difficulty;
  } else {
    bar.length += (barLength - bar.length) * 0.1;
  }
  bar.angle += difficulty * 0.05;
  // Draw a line connecting the coordinates of
  // the first argument and the second argument.
  line(bar.pin, vec(bar.pin).addWithAngle(bar.angle, bar.length));
  if (bar.pin.y > 98) {
    play("explosion");
    // Call 'end()' to end the game. (Game Over)
    end();
  }
  let nextPin;
  // 'remove()' passes the elements of the array of the first argument to
  // the function of the second argument in order and executes it.
  // If the function returns true, the element will be removed from the array.
  remove(pins, (p) => {
    p.y += scr;
    // Draw a box and check if it collides with other black rectangles or lines.
    if (box(p, 3).isColliding.rect.black && p !== bar.pin) {
      nextPin = p;
    }
    return p.y > 102;
  });
  if (nextPin != null) {
    play("powerUp");
    // Add up the score.
    // By specifying the coordinates as the second argument,
    // the added score is displayed on the screen.
    addScore(ceil(bar.pin.distanceTo(nextPin)), nextPin);
    bar.pin = nextPin;
    bar.length = barLength;
  }
  nextPinDist -= scr;
  while (nextPinDist < 0) {
    // 'rnd()' returns a random value.
    pins.push(vec(rnd(10, 90), -2 - nextPinDist));
    nextPinDist += rnd(5, 15);
  }
}
